### ISSUE TITLE
Determine how, or if, Composer should handle identities and multiple enrollments

### ISSUE THREAD

**Issue by davidkel** [hyperledger/composer#1761]
As quoted on RC
```
xuanyue202 12:00 PM 
@sstone1 Hi Simon, I read your new design about the identity and think there might be a problem 
for the current implementation. Currently hash(PEM.raw) is used as the Identifier. Such design
 fails to capture the use-case where user re-enrolls a certificate (the old one get expired) or enrolls
 multiple certificates at the same time (maxEnrollments!=1) and want to use a single identity to
 represent himself. Each time CA issues a new cert, the fields of Serial Number and potentially
 Valid From/To will change, which makes hash(PEM.raw) a different value. I am not sure why we do
 not simply use hash(Common Name, Issuer) as the identifier?
```

Need to check that multiple enrollment of the same identity results in slightly different certificates as claimed above, and need to understand the use case for expired identities. However it was also briefly discussed on the community call where a suggestion to use the SKI of a certificate (for example as used by the fabric-node-sdk) would be a more robust way.

**Comment by davidkel** [comment]
Confirmed that multiple enrollment of the same identity does result in differing certificates. I have also confirmed that the generated SKI for these certificates also differ so you cannot use the SKI as a way to compare if 2 identities are the same.

**Comment by rahianeja** [comment]
Is there a way to bypass this as of now and REST calls to work in multi user setup? Since creating an identity would need participant to exist, and creation of first participant doesn't look possible in secure mode from REST, this looks like a deadlock for me to proceed with multi-user setup.

Can this bug be fixed soon?

**Comment by sstone1** [comment]
@rahianeja this is not a bug, this is working as designed.

We see a lot of misunderstanding around the enrollment secret - it might not be the case here - but in general, a lot of users end up treating the generated enrollment secret as a password. The enrollment secret is a token that you can pass to the Fabric CA in exchange for an enrollment certificate. The enrollment certificate is used to sign transactions and is used to identify you on the Blockchain.

We need to guide our users towards proper management of their enrollment certificates, and we will do that by restricting the default number of maximum enrollments to 1 in issue #2056. This was the behaviour in Fabric v0.6, and the behaviour was much more indicative as to what was going on under the covers. I am asking @jt-nti to add a warning message in playground around this behaviour as well.

After users have enrolled - exchanged the enrollment secret for an enrollment certificate - they will need to manage that enrollment certificate and move it between the wallets (via identity export & import APIs and CLIs we are building) in order to use it across multiple systems.

**Comment by sstone1** [comment]
I am also closing this issue as per the above comments, and the below security considerations.

Every existing system today that performs certificate based authentication stores a hash or fingerprint of the certificate contents; in some cases the entire certificate, or in some cases just the subject public key information. However, the behaviour is always the same - if you lose the certificate and need to generate a new one - you must upload the new certificate into the system after authenticating via some other mechanism. For an example, check out the SSH key management in GitHub.

The `hash(Common Name, Issuer)` we use as the identifier for a newly issued identity is the best we can do in this case. We do not have the certificate because it has not yet been enrolled. We do not want the issuer to perform the enrollment, because then the issuer will have a copy of the private key and can in theory sign transactions as the participant. They still have the enrollment secret, but once we fix issue #2056, the participant will know if somebody else has enrolled using that enrollment secret (because the max enrollment count will be hit).

I am pretty sure that the `hash(Common Name, Issuer)` identifier we use for issued identities is not as secure as security experts would like. There's a reason that nobody else uses it in certificate based authentication. Which is why its currently restricted to newly issued identities, and swiftly replaced with the hash of the certificate contents on activation (first use). In addition, the entire feature can be swiftly disabled with an ACL rule, and users forced into binding existing identities to participants.

If somebody with security expertise wants to do the legwork and research to explain why `hash(Common Name, Issuer)` is actually as safe as what we're doing today, then we can revisit this in the future.
